Refactor of rz-test to use thread safe functions.

wargio opened the pull request:
# do not squash
 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This uses the thread-safe functions and cleans up a bit the rz-test code.

Now we automatically detect the core number and use them all unless `-j <n>` is specified.

wargio committed:
Moved windows static vars into struct.

wargio committed:
Moved unix static vars into struct.

wargio committed:
Refactoring rz-test to use thread safe functions.

Rot127 commented on binrz/rz-test/rz-test.c:
```diff
@@ -26,36 +25,39 @@
+	size_t n_handled; ///< Tests that has been handled
+	size_t ok; ///< Tests that succeeded
+	size_t xx; ///< Tests that failed
+	size_t fx; ///< Tests that succeeded but als marked as broken
```
```suggestion
	size_t fx; ///< Tests that succeeded but also marked as broken
```

XVilka commented on binrz/rz-test/run.c:
```diff
@@ -558,7 +558,7 @@ RZ_API RzTestResultInfo *rz_test_run_test(RzTestRunConfig *config, RzTest *test)
 }
 
-RZ_API void rz_test_test_result_info_free(RzTestResultInfo *result) {
+RZ_API void rz_test_result_info_free(RzTestResultInfo *result) {
```
Add `RZ_NULLABLE` while you are on it

XVilka commented on binrz/rz-test/rz_test.h:
```diff
@@ -213,6 +213,6 @@ RZ_API void rz_test_test_free(RzTest *test);
 RZ_API bool rz_test_test_broken(RzTest *test);
 RZ_API RzTestResultInfo *rz_test_run_test(RzTestRunConfig *config, RzTest *test);
-RZ_API void rz_test_test_result_info_free(RzTestResultInfo *result);
+RZ_API void rz_test_result_info_free(RzTestResultInfo *result);
```
Same

XVilka commented on librz/util/subprocess.c:
```diff
@@ -47,21 +47,25 @@ struct rz_subprocess_t {
+	DWORD mode_stderr;
+} SubprocessWindows;
+
+static SubprocessWindows subwin = { 0 };
```
Add a comment that it is global/static on purpose

XVilka commented on librz/util/subprocess.c:
```diff
@@ -746,28 +750,32 @@ struct rz_subprocess_t {
+	RzThread *sigchld_thread;
+} SubprocessUnix;
+
+static SubprocessUnix subnix = { 0 };
```
Same

XVilka reviewed (changes requested):
Nice cleanup/modernization!

`rz-bindgen` returns error:
```
 [1/7] Generating bindgen_outputs with a custom command
FAILED: rizin.i 
/usr/bin/python3 ../src/main.py -o . --clang-path /usr/lib/llvm-14/lib --clang-args ' -resource-dir="/usr/lib/llvm-14/lib/clang/14.0.0"' --rizin-include-path /usr/include/librz --targets SWIG
Traceback (most recent call last):
  File "/home/runner/work/rizin/rizin/rz-bindgen/build/../src/main.py", line 51, in <module>
    bindings.run()
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/bindings.py", line 43, in run
    builders = [HeaderBuilder(name) for name in threaded_headers]
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/bindings.py", line 43, in <listcomp>
    builders = [HeaderBuilder(name) for name in threaded_headers]
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/cparser_header.py", line 101, in __init__
    assert os.path.exists(filename)
AssertionError
[2/7] Generating plugin/swig_runtime with a custom command
```
The macOS unit test error is now clearer, nice:
```
>> MALLOC_PERTURB_=178 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-2757858299/workspace/build/test/unit/test_core_task
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Hello
this
is
the
main
task!
test_core_task ERR
[XX] Fail at line 44: fcn result: expected 1337, 0
1337, 1
1337, 2
1337, 3
1337, 4
, got (null).

test_core_task_finished_cb ERR
[XX] Fail at line 73: task joined: expected true, got false
```

wargio commented:
The unit tests fails are unrelated to this PR. check dev

XVilka commented:
> The unit tests fails are unrelated to this PR. check dev

I know they are unrelated, I thought the error improved after this one, but it apparently happended in dev already. So yes, ignore that part of the comment.

XVilka commented:
I think having overall count is useful though, could you please also add a column with all tests ran?

Also maybe a final report, how many OK, BR, XX are at the end total?

wargio commented:
> I think having overall count is useful though, could you please also add a column with all tests ran?

i have removed it by mistake. i changed the code and now it prints it.

wargio commented:
The issue with Bindgen is unrelated to this PR
```
AssertionError: File /usr/include/librz/sdb/ls.h does not exist
```

XVilka commented:
> The issue with Bindgen is unrelated to this PR
> 
> ```
> AssertionError: File /usr/include/librz/sdb/ls.h does not exist
> ```

https://github.com/rizinorg/rz-bindgen/pull/64

XVilka reviewed (approved):
There is still room for improvement, e.g. make all XX red in the output if not zero, BR - yellow if not zero, etc. But this is for separate PR.

kazarmy reviewed (commented):
A running total would be nice -- I generally use it to see how far the build has progressed through the tests. Not sure whether it's strictly necessary, or how to integrate that info in a clean manner.

wargio commented:
> A running total would be nice -- I generally use it to see how far the build has progressed through the tests. Not sure whether it's strictly necessary, or how to integrate that info in a clean manner.

The running total is printed at the end when `-L` is used.
```
[21179/21179]    20129 OK      1016 BR        0 XX       34 FX
```
